README
String::UTF8(3)       User Contributed Perl Documentation      String::UTF8(3)



NNAAMMEE
       String::UTF8 − Determine whether a string consists of well−formed UTF−8
       byte sequences

SSYYNNOOPPSSIISS
         use String::UTF8 qw[:all];

         $boolean = is_utf8($string);
         $boolean = is_utf8($string, UTF8_DISALLOW_NONCHARACTERS);

DDEESSCCRRIIPPTTIIOONN
       Determine whether a string consists of well‐formed UTF−8 byte
       sequences.

FFUUNNCCTTIIOONNSS
       iiss__uuttff88

       Determine whether $string consists of well‐formed UTF−8 byte sequences.

       _U_s_a_g_e

           $boolean = is_utf8($string);
           $boolean = is_uft8($string, $flags);

       _A_r_g_u_m_e_n_t_s

       $string
       $flags (optional)

       _R_e_t_u_r_n_s

       $boolean

       _N_o_t_e

       This function ignores Perl’s internal UTF−8 flag ("SVf_UTF8") and sim‐
       ply checks the content of the string.

CCOONNSSTTAANNTTSS
       FFLLAAGGSS


       UTF8_DISALLOW_NONCHARACTERS
           If this flag is set, noncharacters is not considered well−formed.

EEXXPPOORRTT
       None by default. Functions and constants can either be imported indi‐
       vidually or in sets grouped by tag names. The tag names are:

       ":all" exports all functions and constants.

DDIIAAGGNNOOSSTTIICCSS
       ((FF)) Usage: %s
           Subroutine %s invoked with wrong number of arguments.

EENNVVIIRROONNMMEENNTT
       Set the environment variable "STRING_UTF8_PP" to a true value before
       loading this package to disable usage of XS implementation.

PPRREERREEQQUUIISSIITTEESS
       RRuunn‐‐TTiimmee


       perl 5.006 or greater.
       Carp, core module.
       Exporter, core module.

       BBuuiilldd‐‐TTiimmee

       In addition to Run−Time:

       Test::More 0.47 or greater, core module since 5.6.2.
       Test::Exception.

SSEEEE AALLSSOO
       String::UTF8::XS

AAUUTTHHOORR
       Christian Hansen, <chansen@cpan.org>

CCOOPPYYRRIIGGHHTT AANNDD LLIICCEENNSSEE
       Copyright (C) 2009 by Christian Hansen

       This library is free software; you can redistribute it and/or modify it
       under the same terms as Perl itself.



perl v5.8.9                       2009‐12‐01                   String::UTF8(3)


Changes
0.1 2009-12-01 18:00
    - Initial release.



t/lib/myconfig.pm
package myconfig;

use strict;

BEGIN {
    $ENV{STRING_UTF8_PP} = 1;
}

1;


lib/String/UTF8/PP.pm
package String::UTF8::PP;
use strict;
use warnings;

use Carp qw[croak];

BEGIN {
    our $VERSION   = 0.1;
    our @EXPORT_OK = qw(
        is_utf8
        UTF8_DISALLOW_NONCHARACTERS
    );

    require Exporter;
    *import = \&Exporter::import;
}

sub UTF8_DISALLOW_NONCHARACTERS () { 1 }

{
    use bytes;

    my $UTF8_1              = qr/[\x00-\x7F]/;
    my $UTF8_2              = qr/ [\xC2-\xDF] [\x80-\xBF] /x;
    my $UTF8_3              = qr/  \xE0       [\xA0-\xBF] [\x80-\xBF]
                                | [\xE1-\xEC] [\x80-\xBF] [\x80-\xBF]
                                |  \xED       [\x80-\x9F] [\x80-\xBF]
                                | [\xEE-\xEF] [\x80-\xBF] [\x80-\xBF]
                              /x;
    my $UTF8_4              = qr/  \xF0       [\x90-\xBF] [\x80-\xBF] [\x80-\xBF]
                                | [\xF1-\xF3] [\x80-\xBF] [\x80-\xBF] [\x80-\xBF]
                                |  \xF4       [\x80-\x8F] [\x80-\xBF] [\x80-\xBF]
                              /x;

    my $UTF8_3_nonchar      = qr/ \xEF \xB7 [\x90-\xAF]
                                | \xEF \xBF [\xBE-\xBF]
                              /x;

    my $UTF8_4_nonchar      = qr/  \xF0       [\x9F\xAF\xBF]     \xBF [\xBE-\xBF]
                                | [\xF1-\xF3] [\x8F\x9F\xAF\xBF] \xBF [\xBE-\xBF]
                                |  \xF4       [\x8F]             \xBF [\xBE-\xBF]
                              /x;

#   my $UTF8_3_interchange  = qr/ $UTF8_3 (?<!$UTF8_3_nonchar) /x;
    my $UTF8_3_interchange  = qr/  \xE0       [\xA0-\xBF] [\x80-\xBF]
                                | [\xE1-\xEC] [\x80-\xBF] [\x80-\xBF]
                                |  \xED       [\x80-\x9F] [\x80-\xBF]
                                |  \xEE       [\x80-\xBF] [\x80-\xBF]
                                |  \xEF (?:
                                            [\x80-\xB6\xB8-\xBE] [\x80-\xBF]
                                          |  \xB7                [\x80-\x8F\xB0-\xBF]
                                          |  \xBF                [\x80-\xBD]
                                         )
                              /x;

#   my $UTF8_4_interchange  = qr/ $UTF8_4 (?<!$UTF8_4_nonchar) /x;
    my $UTF8_4_interchange  = qr/  \xF0       (?: [\x90-\x9E\xA0-\xAE\xB0-\xBE]             [\x80-\xBF] [\x80-\xBF]
                                                | [\x9F\xAF\xBF]                        (?: [\x80-\xBE] [\x80-\xBF]
                                                                                          |  \xBF       [\x80-\xBD]))
                                | [\xF1-\xF3] (?: [\x80-\x8E\x90-\x9E\xA0-\xAE\xB0-\xBE]    [\x80-\xBF] [\x80-\xBF]
                                                | [\x8F\x9F\xAF\xBF]                    (?: [\x80-\xBE] [\x80-\xBF]
                                                                                          |  \xBF       [\x80-\xBD]))
                                |  \xF4       (?: [\x80-\x8E]                               [\x80-\xBF] [\x80-\xBF]
                                                |  \x8F                                 (?: [\x80-\xBE] [\x80-\xBF]
                                                                                          |  \xBF       [\x80-\xBD]))
                              /x;

    my $UTF8                = qr/ $UTF8_1 | $UTF8_2 | $UTF8_3 | $UTF8_4 /x;
    my $UTF8_interchange    = qr/ $UTF8_1 | $UTF8_2 | $UTF8_3_interchange | $UTF8_4_interchange /x;
    my $UTF8_noncharacter   = qr/ $UTF8_3_nonchar | $UTF8_4_nonchar /x;
    my $UTF8_BOM            = qr/ \xEF \xBB \xBF /x;

    sub is_utf8 {
        @_ == 1 || @_ == 2 || croak(q/Usage: is_utf8(string [, flags])/);
        local $_ = $_[0];
        return ( defined 
              && ((@_ == 1 || !($_[1] & UTF8_DISALLOW_NONCHARACTERS))
                     ? /\A $UTF8* \z/xo
                     : /\A $UTF8_interchange* \z/xo ));
    }
}

1;



lib/String/UTF8.pm
package String::UTF8;
use strict;
use warnings;

BEGIN {
    our $VERSION   = 0.1;
    our @EXPORT_OK = qw(
        is_utf8
        UTF8_DISALLOW_NONCHARACTERS
    );

    our %EXPORT_TAGS = (
        all => [ @EXPORT_OK ],
    );

    if ( !$ENV{STRING_UTF8_PP} ) {

        eval {
            require String::UTF8::XS;
        };
    }

    if ( $ENV{STRING_UTF8_PP} || $@ ) {
        require String::UTF8::PP;
        String::UTF8::PP->import(@EXPORT_OK);
    }
    else {
        String::UTF8::XS->import(@EXPORT_OK);
    }

    require Exporter;
    *import = \&Exporter::import;
}

1;

__END__

=head1 NAME

String::UTF8 - Determine whether a string consists of well-formed UTF-8 byte sequences

=head1 SYNOPSIS

  use String::UTF8 qw[:all];
  
  $boolean = is_utf8($string);
  $boolean = is_utf8($string, UTF8_DISALLOW_NONCHARACTERS);

=head1 DESCRIPTION

Determine whether a string consists of well-formed UTF-8 byte sequences.

=head1 FUNCTIONS

=head2 is_utf8

Determine whether C<$string> consists of well-formed UTF-8 byte sequences.

I<Usage>

    $boolean = is_utf8($string);
    $boolean = is_uft8($string, $flags);

I<Arguments>

=over 4

=item C<$string>

=item C<$flags> (optional)

=back

I<Returns>

=over 4

=item C<$boolean>

=back

I<Note>

This function ignores Perl's internal UTF-8 flag (C<SVf_UTF8>) and simply 
checks the content of the string.

=head1 CONSTANTS

=head2 FLAGS

=over 4

=item UTF8_DISALLOW_NONCHARACTERS

If this flag is set, noncharacters is not considered well-formed.

=back

=head1 EXPORT

None by default. Functions and constants can either be imported individually or
in sets grouped by tag names. The tag names are:

=over 4

=item C<:all> exports all functions and constants.

=back

=head1 DIAGNOSTICS

=over 4

=item B<(F)> Usage: %s

Subroutine %s invoked with wrong number of arguments.

=back

=head1 ENVIRONMENT

Set the environment variable C<STRING_UTF8_PP> to a true value before loading this package to disable usage of XS implementation.

=head1 PREREQUISITES

=head2 Run-Time

=over 4

=item L<perl> 5.006 or greater.

=item L<Carp>, core module.

=item L<Exporter>, core module.

=back

=head2 Build-Time

In addition to Run-Time:

=over 4

=item L<Test::More> 0.47 or greater, core module since 5.6.2.

=item L<Test::Exception>.

=back

=head1 SEE ALSO

L<String::UTF8::XS>

=head1 AUTHOR

Christian Hansen, E<lt>chansen@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Christian Hansen

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
